internal/app/machined/pkg/controllers/network/operator/wglan.go

```diff
@@ -0,0 +1,276 @@
+	"go.uber.org/zap"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	"inet.af/netaddr"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/controllers/network/operator/wglan"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/nethelpers"
+	"github.com/talos-systems/talos/pkg/resources/network"
+)
+
+// WgLAN implements a network operator for controlling the Wireguard LAN system.
+type WgLAN struct {
+	Config *wglan.Config
+
+	privateKey string
+
+	fwMark     int
+	listenPort uint16
+
+	db *wglan.PeerDB
+
+	logger *zap.Logger
+}
+
+// NewWgLAN creates a Wireguard LAN operator.
+func NewWgLAN(logger *zap.Logger, nodenameFunc wglan.NodenameFunc, linkName string, prefix netaddr.IPPrefix, clusterID string, privateKey string, discoveryURL string) *WgLAN {
+	if discoveryURL == "" {
+		discoveryURL = constants.WireguardDefaultNATDiscoveryService
+	}
+
+	privKey, err := wgtypes.ParseKey(privateKey)
+	if err != nil {
+		logger.Sugar().Fatalf("failed to parse Wireguard private key %q: %w", privateKey, err)
+	}
+
+	ip, err := wgEUI64(prefix)
+	if err != nil {
+		logger.Warn("failed to generate local IP address", zap.Error(err))
+
+		return nil
+	}
+
+	pubKey := privKey.PublicKey().String()
+
+	cfg := &wglan.Config{
+		ClusterID:    clusterID,
+		DiscoveryURL: discoveryURL,
+		IP:           ip,
+		LinkName:     linkName,
+		Nodename:     nodenameFunc,
+		PublicKey:    pubKey,
+		RoutingTable: constants.WireguardDefaultRoutingTable,
+		Subnet:       prefix,
+	}
+
+	return &WgLAN{
+		Config:     cfg,
+		logger:     logger,
+		fwMark:     constants.WireguardDefaultFirewallMark,
+		privateKey: privateKey,
+		listenPort: constants.WireguardDefaultPort,
+		db:         new(wglan.PeerDB),
+	}
+}
+
+// Prefix returns unique operator prefix which gets prepended to each spec.
+func (o *WgLAN) Prefix() string {
+	return fmt.Sprintf("wglan/%s", o.Config.LinkName)
+}
+
+// Run the operator loop.
+func (o *WgLAN) Run(ctx context.Context, notifyCh chan<- struct{}) {
+	var rulesManager *wglan.RulesManager
+
+	var peerManager *wglan.PeerManager
+
+	for ctx.Err() == nil {
+		time.Sleep(time.Second)
+
+		rulesManager = new(wglan.RulesManager)
+
+		if err := rulesManager.Run(ctx, o.logger, o.db); err != nil {
+			o.logger.Warn("failed to start rules manager", zap.Error(err))
+
+			continue
+		}
+
+		peerManager = wglan.NewPeerManager(o.Config, o.db)
+
+		if err := peerManager.Run(ctx, o.logger, notifyCh); err != nil {
+			o.logger.Warn("failed to start peer manager", zap.Error(err))
+		}
+	}
+}
+
+// AddressSpecs implements Operator interface.
+func (o *WgLAN) AddressSpecs() []network.AddressSpecSpec {
+	if o == nil || o.Config == nil {
+		return nil
+	}
+
+	return []network.AddressSpecSpec{
+		{
+			Address:         o.Config.IP,
+			LinkName:        o.Config.LinkName,
+			Family:          unix.AF_INET6,
+			Scope:           unix.RT_SCOPE_UNIVERSE,
+			Flags:           0,
```

Comment:
we should have `permanent` flag here, as otherwise it is set by the kernel and address starts flapping